bandersnatch/bandersnatch.go

```diff
@@ -92,3 +96,62 @@ func computeY(x *fp.Element, choose_largest bool) *fp.Element {
 		return sqrtY.Neg(sqrtY)
 	}
 }
+
+// PointExtendedFromProj converts a point in projective coordinates to extended coordinates.
+func PointExtendedFromProj(p *PointProj) PointExtended {
+	var pzinv fp.Element
+	pzinv.Inverse(&p.Z)
+	var z fp.Element
+	z.Mul(&p.X, &p.Y).Mul(&z, &pzinv)
+	return PointExtended{
+		X: p.X,
+		Y: p.Y,
+		Z: p.Z,
+		T: z,
+	}
+}
+
+// PointExtendedNormalized is an extended point which is normalized.
+// i.e: Z=1. We store it this way to save 32 bytes per point in memory.
+type PointExtendedNormalized struct {
+	X, Y, T gnarkfr.Element
+}
+
+// Neg computes p = -p1
+func (p *PointExtendedNormalized) Neg(p1 *PointExtendedNormalized) *PointExtendedNormalized {
+	p.X.Neg(&p1.X)
+	p.Y = p1.Y
+	p.T.Neg(&p1.T)
+	return p
+}
+
+// ExtendedAddNormalized computes p = p1 + p2.
+// https://hyperelliptic.org/EFD/g1p/auto-twisted-extended.html#addition-madd-2008-hwcd
+func ExtendedAddNormalized(p, p1 *PointExtended, p2 *PointExtendedNormalized) *gnarkbandersnatch.PointExtended {
+	var A, B, C, D, E, F, G, H, tmp gnarkfr.Element
+	A.Mul(&p1.X, &p2.X)
+	B.Mul(&p1.Y, &p2.Y)
+	C.Mul(&p1.T, &p2.T).Mul(&C, &CurveParams.D)
+	D.Set(&p1.Z)
+	tmp.Add(&p1.X, &p1.Y)
+	E.Add(&p2.X, &p2.Y).
+		Mul(&E, &tmp).
+		Sub(&E, &A).
+		Sub(&E, &B)
+	F.Sub(&D, &C)
+	G.Add(&D, &C)
+	H.Set(&A)
+
+	// mulBy5(&H)
+	H.Neg(&H)
+	gnarkfr.MulBy5(&H)
```

Comment:
gnark-crypto uses `mulBy5(..)` but that's unexported; but it's equivalent to L146+L147